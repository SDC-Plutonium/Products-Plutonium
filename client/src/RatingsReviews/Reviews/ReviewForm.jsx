import React from 'react';
import styled from 'styled-components';

const StyledForm = styled.div`
  display: flex; /* Hidden by default */
  flex-direction: column;
  align-items: center;
  justify-content: center;
  position: absolute; /* Stay in place */
  z-index: 1; /* Sit on top */
  left: 0;
  top: 0;
  width: 100%; /* Full width */
  height: 100%; /* Full height */
  overflow: auto;
  background-color: rgb(0,0,0); /* Fallback color */
  background-color: rgba(0,0,0,0.4); /* Black w/ opacity */
`

const StyledInner = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 80%;
  padding: 10%;
`

const StyledInput = styled.input`
  width: 50%;
`

const StyledTextArea = styled.textarea`
  width: 300px;
  height: 120px;
  resize: none;
`
const StyledClose = styled.button`
  color: #aaa;
  float: right;
  font-size: 28px;
  font-weight: bold;
`

class ReviewForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      rating: '',
      summary: '',
      body: '',
      name: '',
      email: '',
      recommend: false,
    };
    this.addReview = this.addReview.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.recommend = this.recommend.bind(this);
    this.closeForm = this.closeForm.bind(this);
  }

  closeForm() {

  }

  addReview(e) {
    // format it here
    // this.props.addReview();
    e.preventDefault();
    const reviewBody = {
      product_id: 37311,
      rating: parseInt(this.state.rating),
      summary: this.state.summary,
      body: this.state.body,
      name: this.state.name,
      email: this.state.email,
      recommend: this.state.recommend,
      characteristics: {'125033': 3, '125031': 4, '125032': 5, '125034': 3}
    };
    console.log(reviewBody);
    this.props.addReview(reviewBody);
  }

  handleChange(e) {
    e.preventDefault();
    this.setState({
      [e.target.name]: e.target.value,
    });
  }

  recommend(e) {
    e.preventDefault();
    this.setState({ recommend: true });
  }

  render() {
    return (
      <StyledForm>
        <h2>here is the review form.</h2>
        <StyledInner>

            <StyledInput placeholder="rating" name="rating" onChange={this.handleChange}/>
            <textarea placeholder="summary" name="summary" onChange={this.handleChange} />
            <textarea placeholder="body" name="body" onChange={this.handleChange} />
            <button onClick={this.recommend}>recommend?</button>
            <StyledInput placeholder="name" name="name" onChange={this.handleChange} />
            <StyledInput placeholder="email" name="email" onChange={this.handleChange} />
            <StyledInput type="submit" onClick={this.addReview} />
            <StyledClose onClick={this.closeForm}>CLOSE</StyledClose>

        </StyledInner>

      </StyledForm>
    );
  }
}

export default ReviewForm;
