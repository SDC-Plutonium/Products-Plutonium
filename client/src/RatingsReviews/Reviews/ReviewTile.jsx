/* eslint-disable no-alert */
import React from 'react';
import { format, parseISO } from 'date-fns';
import styled from 'styled-components';

const StyledTile = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: top;
  width: 100%;
  border-top: 2px solid black;
  padding: 5%;
`

const StyledButton = styled.button`
  width: 100px;
  font-size: 1em;
  margin: 1em;
  padding: 0.25em 1em;
  border-radius: 3px;
  background: white;
  color: black;
  border: 1px solid black;
  &:hover {
    cursor: pointer;
    opacity: 60%;
  }
`

class ReviewTile extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      reviewed: [],
    };
    this.markHelpful = this.markHelpful.bind(this);
    this.report = this.report.bind(this);
  }

  markHelpful(e) {
    e.preventDefault();
    let rId = this.props.review.review_id
    this.props.markHelpful(rId);
  };

  report(e) {
    e.preventDefault();
    const r = confirm('are you sure you want to report this review?');
    if (r) {
      this.props.report(this.props.review.review_id);
    }
  }

  render() {
    return (
      <StyledTile>
        Rating: {this.props.review.rating + " "}
        Date: {format(parseISO(this.props.review.date), 'MMMM dd, yyyy') + " "}
        <h5>{this.props.review.summary}</h5>
        {this.props.review.body + " "}
        {this.props.review.recommend &&
          <h5>
            I recommend this product &#10003;
          </h5>
        }
        {this.props.review.reviewer_name &&
          <h5>
            By: {this.props.review.reviewer_name + " "}
          </h5>
        }
        {this.props.review.response &&
          <h4>
            from the seller: {this.props.review.response}
          </h4>
        }
        Helpful? : {this.props.review.helpfulness + " "}
        <div>
          <StyledButton onClick={this.markHelpful}>YES</StyledButton>
          <StyledButton onClick={this.report}>report</StyledButton>
        </div>

      </StyledTile>
    )
  }
}

export default ReviewTile;
