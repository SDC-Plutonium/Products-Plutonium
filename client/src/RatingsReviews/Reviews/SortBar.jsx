import React from 'react';

function SortBar(props) {
  const sort = (e) => {
    props.sort(e.target.value);
  }

  return (
    <div>
      {props.reviews.length} reviews, sorted by
      <select id="sort" onChange={sort}>
        <option value="relevance">Relevance</option>
        <option value="newest">Newest</option>
        <option value="helpful">Helpful</option>
      </select>
    </div>
  )
}
export default SortBar;