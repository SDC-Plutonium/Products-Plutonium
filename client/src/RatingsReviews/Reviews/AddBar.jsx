import React from 'react';
import ReviewForm from './ReviewForm.jsx';

class AddBar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      formShowing: false,
    };
    this.moreReviews = this.moreReviews.bind(this);
    this.toggleForm = this.toggleForm.bind(this);
  }

  moreReviews(e) {
    e.preventDefault();
    this.props.moreReviews();
  }

  toggleForm() {
    this.state.formShowing
      ? this.setState({ formShowing: false })
      : this.setState({ formShowing: true });
  }

  render() {
    return (
      <div>
        {`____________________________`}<br />
        <button onClick={this.moreReviews}>MORE</button>
        <button onClick={this.toggleForm}>ADD REVIEW</button>
        {this.state.formShowing && <ReviewForm addReview={this.props.addReview} />}
      </div>
    );
  }
}

export default AddBar;
