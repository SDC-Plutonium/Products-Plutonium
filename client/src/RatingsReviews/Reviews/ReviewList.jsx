import React from 'react';
import ReviewTile from './ReviewTile.jsx';
import styled from 'styled-components';

const StyledList = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: top;
  width: 100%;
  padding: 5%;
`

class ReviewList extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      reviews: [],
    };
  }

  render() {
    return (this.props.reviews.length > 0) ? (
      <StyledList>
        {this.props.reviews.map((review, index) => (
          <ReviewTile
            review={review}
            key={index}
            markHelpful={this.props.markHelpful}
            report={this.props.report}
          />
        ))}
      </StyledList>
    ) : (
      <StyledList>
        <h2>
          There are no reviews currently
        </h2>
      </StyledList>
    );
  }
}

export default ReviewList;
