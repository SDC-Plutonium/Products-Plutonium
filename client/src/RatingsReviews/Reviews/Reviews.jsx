/* eslint-disable class-methods-use-this */
import React from 'react';
import axios from 'axios';
import ReviewList from './ReviewList.jsx';
import AddBar from './AddBar.jsx';
import SortBar from './SortBar.jsx';
import styled from 'styled-components';

const StyledReviews = styled.div`
  display: flex;
  flex-direction: column;
  align-items: top;
  width: 90%;
  padding: 5%;

`

class Reviews extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      reviews: [],
      page: 1,
      markedHelpful: []
    };
    this.getReviews = this.getReviews.bind(this);
    this.addReview = this.addReview.bind(this);
    this.moreReviews = this.moreReviews.bind(this);
    this.sort = this.sort.bind(this);
    this.markHelpful = this.markHelpful.bind(this);
    this.report = this.report.bind(this);
    this.addReview = this.addReview.bind(this);
    this.refresh = this.refresh.bind(this);
  }

  componentDidMount() {
    this.getReviews();
  }

  getReviews() {
    axios.get(`${process.env.API_URL}/reviews?product_id=37311&count=${this.state.page * 2}`, {
      headers: {
        Authorization: process.env.AUTH_KEY,
      },
    })
      .then((response) => {
        console.log('successfully fetched reviews');
        console.log(response.data.results);
        this.setState(
          { reviews: response.data.results },
        );
      })
      .catch((err) => console.log('error fetching reviews', err));
  }

  moreReviews() {
    let page = this.state.page;
    page += 1;
    this.setState({ page });
    this.getReviews();
  }

  sort(option) {
    axios.get(`${process.env.API_URL}/reviews?product_id=37311&sort=${option}`, {
      headers: {
        Authorization: process.env.AUTH_KEY,
      },
    })
      .then((response) => {
        console.log('successfully fetched reviews');
        console.log(response.data.results);
        this.setState(
          { reviews: response.data.results },
        );
      })
      .catch((err) => console.log('error fetching reviews', err));
  }

  addReview(reviewBody) {
    // let temp = {
    //   product_id: 37311,
    //   rating: 5,
    //   name: 'joe',
    //   summary: 'wow',
    //   body: 'cool',
    //   recommend: true,
    //   email: 'aaa@aaa.com',
    //   characteristics: {'125033': 3, '125031': 4, '125032': 5, '125034': 3}
    // };
    // console.log(temp);

    axios.post(`${process.env.API_URL}/reviews`, reviewBody, {
      headers: {
        AUthorization: process.env.AUTH_KEY,
      },
    })
      .then((response) => {
        console.log('success adding review', response);
        alert('thank you for your submission');
        // this.getReviews();
        this.refresh();
      })
      .catch((err) => console.log('error adding review', err));
  }

  refresh() {
    window.location.reload();
  }

  report(reviewId) {
    axios.put(`${process.env.API_URL}/reviews/${reviewId}/report`, {review_id: reviewId}, {
      headers: {
        Authorization: process.env.AUTH_KEY,
      },
    })
      .then(() => {
        this.setState({ reviewed: true });
        console.log('success reporting review');
        this.getReviews();
      })
      .catch((err) => console.log('error reporting review', err))
  }

  markHelpful(reviewId) {
    if (this.state.markedHelpful.indexOf(reviewId) === -1) {
      axios.put(`${process.env.API_URL}/reviews/${reviewId}/helpful`, {review_id: reviewId}, {
        headers: {
          Authorization: process.env.AUTH_KEY,
        },
      })
        .then(() => {
          let markedHelpful = this.state.markedHelpful;
          markedHelpful.push(reviewId);
          this.setState({ markedHelpful });
          this.getReviews();
        })
        .catch((err) => console.log('error marking helpful', err))
    } else {
      alert('you have already marked this review as helpful');
    }
  }

  render() {
    return (
      <div data-testid="reviews-1">
        <h5>R E V I E W S .</h5>
        <SortBar
          reviews={this.state.reviews}
          sort={this.sort}
        />
        <ReviewList
          reviews={this.state.reviews}
          markHelpful={this.markHelpful}
          report={this.report}
        />
        <AddBar
          reviews={this.state.reviews}
          addReview={this.addReview}
          moreReviews={this.moreReviews}
        />
      </div>
    )
  }
}

export default Reviews;
