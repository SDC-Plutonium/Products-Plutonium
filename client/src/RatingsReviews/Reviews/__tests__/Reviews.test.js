import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, cleanup } from '@testing-library/react';
import Reviews from '../Reviews';

test('should render Reviews component', () => {
  const ReviewsElement = renderer.create(<Reviews />);
  expect(typeof ReviewsElement).toEqual('object');
})

test('renders correctly', () => {
  const tree = renderer
    .create(<Reviews />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

