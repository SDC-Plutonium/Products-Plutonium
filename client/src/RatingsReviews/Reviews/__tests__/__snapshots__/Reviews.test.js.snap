// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  data-testid="reviews-1"
>
  <h5>
    R E V I E W S .
  </h5>
  <div>
    0
     reviews, sorted by
    <select
      id="sort"
      onChange={[Function]}
    >
      <option
        value="relevance"
      >
        Relevance
      </option>
      <option
        value="newest"
      >
        Newest
      </option>
      <option
        value="helpful"
      >
        Helpful
      </option>
    </select>
  </div>
  <div
    className="sc-dkzDqf dMbqVG"
  >
    <h2>
      There are no reviews currently
    </h2>
  </div>
  <div>
    ____________________________
    <br />
    <button
      onClick={[Function]}
    >
      MORE
    </button>
    <button
      onClick={[Function]}
    >
      ADD REVIEW
    </button>
  </div>
</div>
`;
